grove/amplification/amplification.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2016-2017 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+
+"""
+Module for amplitude amplification, for use in algorithms such as Grover's
+algorithm.
+
+For more information, see arXiv:quant-ph/0005055.
+"""
+
+import numpy as np
+import pyquil.quil as pq
+from pyquil.gates import H, X, Z, RZ, STANDARD_GATES
+from scipy.linalg import sqrtm
+
+STANDARD_GATE_NAMES = STANDARD_GATES.keys()
+
+
+# Start with A|0>
+
+def amplify(A, A_inv, U_w, qubits, num_iter, init=True):
+    """
+    Returns a program that does n rounds of amplification,
+    given a measurement-less algorithm A, an oracle U_w,
+    and a list of qubits to operate on.
+
+    :param A: a program representing a measurement-less algorithm run on qubits
+    :param A_inv: a program representing the inverse algorithm of A.
+                  This can be done using the Program
+                  object's adjoint() method
+    :param U_w: an oracle maps any basis vector to either |0> or |1>
+    :param qubits: the qubits to operate on
+    :param num_iter: number of iterations of amplifications to run
+    :param init: a boolean flag that is set to True if and only if A
+                 is to be applied initially on the input qubits.
+                 By default, it is set to True.
+    :return:
+    """
+
+    # Assertions to check input
+    assert isinstance(A, pq.Program), \
+        "A must be a valid Program instance"
+    assert isinstance(A_inv, pq.Program), \
+        "A_inv must be a valid Program instance"
+    assert isinstance(U_w, pq.Program), \
+        "U_w must be a valid Program instance"
+    assert num_iter > 0, \
+        "The number of iterations must be greater than 0"
+    assert len(qubits) > 0, \
+        "The list of qubits to apply the diffusion " \
+        "operator to must be non-empty"
+
+    p = A if init else pq.Program()
+
+    for _ in xrange(num_iter):
+        # A (2|0><0| - I) A^-1 (I - 2|w><w|) n times
+        p += U_w + A_inv + diffusion_operator(qubits) + A
+
+    return p
+
+
+def n_qubit_control(controls, target, u, gate_name):
+    """
+    Returns a controlled u gate with n-1 controls.
+    Useful for constructing oracles.
+
+    Uses a number of gates quadratic in the number of qubits,
+    and defines a linear number of new gates. (Roots and adjoints of u.)
```

Comment:
Could you provide a pointer (whether in the documentation or in the source) to where you learned this algorithmic construction from? That would be useful in an example module. (It looks to me like the given reference quant-ph/0005055 assumes you already know how to construct these Lambda gates.)